### PR TITLE
Add CollaborativeOptimizer

### PR DESCRIPTION
**Pull request by leshanbog**
This PR introduces necessary objects for collaborative training so that client code for participation is clean and concise (like default train loop).
- WeightedAverager - DecentralizedAverager that averages trainable params or gradients with peer-wise weights
- CollaborativeOptimizer - performs model updates after collaboratively accumulating a target (large) batch size across peers

### CODE REVIEW

**Comment by justheuristic** [comment]
[example usage]
```
import time, socket
import torch, torch.nn as nn
import hivemind

with socket.socket() as sock:
    coordinator_exists = sock.connect_ex(("127.0.0.1", 1337)) == 0
if not coordinator_exists:
    dht = hivemind.DHT(listen_on='127.0.0.1:1337', start=True),
    
model = nn.Sequential(nn.Linear(32, 16), nn.ReLU(), nn.Linear(16, 32))

opt = hivemind.CollaborativeOptimizer(
    opt=torch.optim.Adam(model.parameters()),
    dht=hivemind.DHT(initial_peers=['127.0.0.1:1337'], start=True),
    prefix='test_exp', target_group_size=2,
    target_batch_size=32, batch_size_per_step=1,
    start=True
)


while True:
    x = torch.randn(10, 32)
    time.sleep(1)
    loss = torch.mean((x - model(x)) ** 2)
    loss.backward()
    opt.step()
```

File: hivemind/client/optim/performance_ema.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        :param num_processed: how many items were processed since last call
+        :returns: current estimate of performance (samples per second), but at most
+        """
+        assert not self.paused, "PerformanceEMA is currently paused"
+        assert num_processed > 0, f"Can't register processing {num_processed} samples"
+        self.timestamp, old_timestamp = get_dht_time(), self.timestamp
+        seconds_per_sample = max(0, self.timestamp - old_timestamp) / num_processed
+        self.ema_seconds_per_sample = self.alpha * seconds_per_sample + (1 - self.alpha) * self.ema_seconds_per_sample
+        self.num_updates += 1
+        adjusted_seconds_per_sample = self.ema_seconds_per_sample / (1 - (1 - self.alpha) ** self.num_updates)
```
Comment (mryab) [inline hivemind/client/optim/performance_ema.py:28]: You don't need this adjustment if you're already computing a convex combination at each step, see the [link](https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average)

File: hivemind/client/optim/collaborative.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+logger = get_logger(__name__)
+LRSchedulerBase = getattr(torch.optim.lr_scheduler, '_LRScheduler', None)
+
+#TODO make sure that the weighted averager will NOT advertize for averaging if it is out of sync!
+# warn if expiration is too infrequent (measure mean time between steps)
+#TODO do not forget to setup client mode!
+#TODO make sure we can recover from zero_grad!
+#TODO actually use scheduler normally on step!
+#TODO enforce minimum period on reporting progress
```
Comment (mryab) [inline hivemind/client/optim/collaborative.py:24]: Either resolve this before merging the PR or create corresponding issues

File: hivemind/client/optim/collaborative.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+                    f"Skipped averaging: collaboration consists of {self.collaboration_state.num_peers} peer(s).")
+                self.averager.local_step += 1
+
+            output = self.opt.step()
+            self.opt.zero_grad()
+            self.local_samples_accumulated = self.local_steps_accumulated = 0
+            self.collaboration_state.register_step()
+            self.state_updated.set()
+            self.update_scheduler()
+            logger.info(f"Optimizer step: done!")
```
Comment (mryab) [inline hivemind/client/optim/collaborative.py:164]: debug?

File: hivemind/client/optim/collaborative.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+            return
+
+        self.local_samples_accumulated += batch_size
+        self.local_steps_accumulated += 1
+        self.performance_ema.update(num_processed=self.batch_size_per_step)
+        run_in_background(self.report_training_progress)
+        if not self.collaboration_state.ready_for_step:
+            return
+
+        logger.info("Averaging parameters and gradients with peers...")
```
Comment (mryab) [inline hivemind/client/optim/collaborative.py:140]: debug?

File: hivemind/client/optim/collaborative.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+                mean_samples_per_worker = self.target_batch_size / self.collaboration_state.num_peers
+                weight = self.local_samples_accumulated / mean_samples_per_worker
+                self.averager.step(weight=weight, timeout=self.averaging_timeout, **kwargs)
+            else:
+                logger.info(
+                    f"Skipped averaging: collaboration consists of {self.collaboration_state.num_peers} peer(s).")
+                self.averager.local_step += 1
+
+            output = self.opt.step()
+            self.opt.zero_grad()
```
Comment (mryab) [inline hivemind/client/optim/collaborative.py:159]: Is it true that we ignore all gradient accumulation and LR scheduling in external code? If so, we should make it very clear in the docstring

File: hivemind/client/optim/collaborative.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+            # note: we deliberately count only valid peers for samples_accumulated, but all peers for performance;
+            # the rationale behind this is that outdated peers will synchronize and begin contributing shortly.
+
+        estimated_samples_remaining = self.target_batch_size - estimated_curent_samples
+        estimated_time_to_next_step = max(0, estimated_samples_remaining) / total_samples_per_second
+
+        expected_max_peers = max(num_peers + self.expected_drift_peers, num_peers * (1 + self.expected_drift_rate))
+        time_to_next_fetch = float(np.clip(a=estimated_time_to_next_step * num_peers / expected_max_peers,
+                                           a_min=self.min_refresh_period, a_max=self.max_refresh_period))
+        logger.info(f"Collaboration accumulated {total_samples_accumulated} samples from {num_peers} peers; "
```
Comment (mryab) [inline hivemind/client/optim/collaborative.py:230]: debug?

File: hivemind/client/optim/collaborative.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        return isinstance(state, (list, tuple)) and len(state) == 5 \
+               and all(map(isinstance, state, (int, int, float, float, bool)))
+
+    def update_scheduler(self):
+        if self.scheduler:
+            while self.scheduler._step_count < self.local_step:
+                self.scheduler.step()
+
+    def shutdown(self):
+        logger.info("Shutting down averager...")
```
Comment (mryab) [inline hivemind/client/optim/collaborative.py:251]: debug

**Review by mryab — CHANGES_REQUESTED** [review]
The rationale for all logger.debug replacements is that we don't want to pollute the output of training scripts unless asked by the user. An alternative solution is to control the verbosity by an argument

File: hivemind/client/averaging/weighted.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+                 extra_tensors: Sequence[torch.Tensor] = (), initialize_optimizer: bool = True, **kwargs):
+
+        self.opt, self.extra_tensors, self.local_step = opt, tuple(extra_tensors), 0
+        self.average_parameters, self.average_gradients = average_parameters, average_gradients
+        self.lock_averager_step = Lock()
+        if initialize_optimizer:
+            initialize_optimizer_state(opt)  # note: this will run one optimizer step!
+
+        with torch.no_grad():
+            averaged_tensors = [tensor.detach().cpu().float().clone() for tensor in self.local_tensors()]
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/client/averaging/weighted.py:40]: 1. yep, see screenshot
image

2. yep

File: hivemind/client/averaging/weighted.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+from threading import Lock
+from typing import Sequence, Dict, Iterator
+import torch
+from hivemind.client.averaging import DecentralizedAverager
+from hivemind.utils import nested_flatten, nested_pack, get_logger, run_in_background
+
+logger = get_logger(__name__)
+
+
+class WeightedAverager(DecentralizedAverager):
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/client/averaging/weighted.py:12]: [will do, as agreed]

[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/client/averaging/weighted.py]: solved by #217

File: hivemind/client/averaging/weighted.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+                group_weights = super().step(gather=weight, **kwargs)
+                logger.info(f"Averaged parameters with {len(group_weights)} peers.")
+                averaging_failed = False
+            except Exception as e:
+                logger.error(f"Averaging step failed: {e}")
+                group_weights = {self.endpoint: weight}
+                averaging_failed = True
+
+            # All averaged parameters were pre-multiplied by the weights of the corresponding peers.
+            # Let us compensate for that by dividing weights by the sum of grad scales over the entire group.
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/client/averaging/weighted.py]: resolved by #217

File: hivemind/client/averaging/weighted.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+            sum_of_weights = sum(weight for weight in group_weights.values() if isinstance(weight, float))
+            normalization_coefficient = (len(group_weights) / sum_of_weights) if sum_of_weights > 0 else 1.0
+
+            with torch.no_grad(), self.get_tensors() as averaged_tensors:
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/client/averaging/weighted.py]: done

File: hivemind/client/optim/performance_ema.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        self.samples_per_second = 1 / max(adjusted_seconds_per_sample, self.eps)
+        return self.samples_per_second
+
+    @contextmanager
+    def pause(self):
+        """ While inside this context, EMA will not count the time passed towards the performance estimate """
+        self.paused, was_paused = True, self.paused
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/client/optim/performance_ema.py:35]: [wontfix, based on previous discussion]

File: hivemind/client/optim/collaborative.py (hunk 4 above)
Comment (justheuristic) [reply hivemind/client/optim/collaborative.py]: yep

File: hivemind/client/optim/collaborative.py (hunk 5 above)
Comment (justheuristic) [reply hivemind/client/optim/collaborative.py:173]: added a better description, will create an issue for a "safe" option

File: hivemind/client/optim/collaborative.py (hunk 2 above)
Comment (justheuristic) [reply hivemind/client/optim/collaborative.py]: done

File: hivemind/client/optim/performance_ema.py (hunk 1 above)
Comment (justheuristic) [reply hivemind/client/optim/performance_ema.py:28]: Linked code: https://github.com/tqdm/tqdm/blob/master/tqdm/std.py#L217-L245

File: hivemind/client/averaging/training.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+import torch
+from hivemind.client.averaging import DecentralizedAverager
+from hivemind.utils import nested_flatten, nested_pack, get_logger, run_in_background
+
+logger = get_logger(__name__)
+
+
+class TrainingAverager(DecentralizedAverager):
+    """
+    An high-level interface to DecentralizedAverager that averages trainable params or gradients for an optimizer.
```
Comment (mryab) [inline hivemind/client/averaging/training.py:14]:
```suggestion
    A high-level interface to DecentralizedAverager that averages trainable params or gradients for an optimizer.
```

File: hivemind/client/optim/collaborative.py [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+        if batch_size is not None and self.batch_size_per_step is None:
+            raise ValueError("please either set batch_size_per_step parameter at init or provide batch_size in .step")
+        batch_size = self.batch_size_per_step if batch_size is None else batch_size
+
+        if not self.is_synchronized:
+            self.load_state_from_peers()
+            return
+
+        if self.last_step_time is not None and get_dht_time() - self.last_step_time > self.metadata_expiration:
+            logger.warning(f"Training step took {get_dht_time() - self.last_step_time}, but metadata {self.metadata_expiration}, but ")
```
Comment (mryab) [inline hivemind/client/optim/collaborative.py:145]: Missing end of the message?

File: hivemind/client/optim/collaborative.py [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+
+    def step(self, batch_size: Optional[int] = None, **kwargs):
+        """
+        Report accumulating gradients w.r.t. batch_size additional samples, optionally update model parameters
+
+        :param batch_size: optional override for batch_size_per_step from init
+        :note: this .step is different from normal pytorch optimizers in several key ways. See __init__ for details.
+        """
+        if batch_size is not None and self.batch_size_per_step is None:
+            raise ValueError("please either set batch_size_per_step parameter at init or provide batch_size in .step")
```
Comment (mryab) [inline hivemind/client/optim/collaborative.py:137]:
```suggestion
            raise ValueError("Please either set batch_size_per_step parameter at init or provide batch_size in .step")
```

File: hivemind/client/optim/collaborative.py (hunk 14 above)
Comment (justheuristic) [reply hivemind/client/optim/collaborative.py:145]: thanks!

File: hivemind/client/optim/collaborative.py [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+
+
+class CollaborativeOptimizer(DecentralizedOptimizerBase):
+    """
+    An optimizer that performs model updates after collaboratively accumulating a target (large) batch size across peers
+
+    These optimizers use DHT to track how much progress did the collaboration make towards target batch size.
+    Once enough samples were accumulated, optimizers will compute a weighted average of their statistics.
+
+    :note: this optimizer behaves unlike regular pytorch optimizers in two ways:
```
Comment (mryab) [inline hivemind/client/optim/collaborative.py:48]:
```suggestion
    :note: This optimizer behaves unlike regular pytorch optimizers in two ways:
```

File: hivemind/client/averaging/training.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
+""" An extension of averager that supports common optimization use cases. """
+from itertools import chain
+from threading import Lock
+from typing import Sequence, Dict, Iterator
+import torch
```
Comment (mryab) [inline hivemind/client/averaging/training.py:5]: Missing empty line

Comment (justheuristic) [reply hivemind/client/averaging/training.py]: thanks, fixed i

**Review by mryab — APPROVED** [review]

File: hivemind/client/optim/collaborative.py [hunk 18]
```diff
@@ -0,0 +1,283 @@
...
+            metadata_expiration: float = 30.0, averaging_timeout: Optional[float] = None, verbose: bool = False,
+            **kwargs):
+        super().__init__(opt, dht)
+        self.prefix, self.scheduler = prefix, scheduler
+        self.target_batch_size, self.batch_size_per_step = target_batch_size, batch_size_per_step
+        self.min_refresh_period, self.max_refresh_period, self.default_refresh_period =\
+            min_refresh_period, max_refresh_period, default_refresh_period
+        self.expected_drift_peers, self.expected_drift_rate = expected_drift_peers, expected_drift_rate
+        self.averaging_timeout, self.metadata_expiration = averaging_timeout, metadata_expiration
+        self.averager = TrainingAverager(
```
Comment (nevec) [inline hivemind/client/optim/collaborative.py:90]: Maybe we should accept averager class in constructor or give a possibility to override averager in subclass?
I don't see an easy way to change averaging logic (e.g. average opt stats) without copypasting the whole class. Correct me if I'm wrong

Comment (justheuristic) [reply hivemind/client/optim/collaborative.py]: Good point, that might prove useful in subsequent research for secure/robust averagers
/* Added averager_cls parameter that allows user to override the averager class */

Comment (justheuristic) [reply hivemind/client/optim/collaborative.py]: Conversation log with @nevec : we have instead opted for a _make_averager method for more versatility
